flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveBatchScheduler.java

```diff
@@ -162,10 +162,7 @@ public AdaptiveBatchScheduler(
         this.vertexParallelismAndInputInfosDecider =
                 checkNotNull(vertexParallelismAndInputInfosDecider);
 
-        this.forwardGroupsByJobVertexId =
-                ForwardGroupComputeUtil.computeForwardGroups(
-                        jobGraph.getVerticesSortedTopologicallyFromSources(),
-                        getExecutionGraph()::getJobVertex);
+        this.forwardGroupsByJobVertexId = checkNotNull(forwardGroupsByJobVertexId);
```

Comment:
Why not keep the forwardGroup creating in `AdaptiveBatchScheduler`?

Reply:
Because we need to ensure that the parallelism of vertices in the forward group has been set when calling `computeVertexParallelismStoreForDynamicGraph`.